clang/lib/AST/Expr.cpp

```diff
@@ -503,6 +503,93 @@ SourceLocation DeclRefExpr::getEndLoc() const {
+  setDependence(computeDependence(this));
+}
+
+UniqueStableNameExpr::UniqueStableNameExpr(EmptyShell Empty, QualType ResultTy,
+                                           bool IsExpr)
+    : Expr(UniqueStableNameExprClass, ResultTy, VK_RValue, OK_Ordinary) {
+  Kind = IsExpr ? ParamKind::Expr : ParamKind::Expr;
+}
+
+UniqueStableNameExpr::UniqueStableNameExpr(SourceLocation OpLoc,
+                                           SourceLocation LParen,
+                                           SourceLocation RParen,
+                                           QualType ResultTy, Expr *E)
+    : Expr(UniqueStableNameExprClass, ResultTy, VK_RValue, OK_Ordinary),
+      OpLoc(OpLoc), LParen(LParen), RParen(RParen), Kind(ParamKind::Expr) {
+  setExpr(E);
+  setDependence(computeDependence(this));
+}
+
+UniqueStableNameExpr *UniqueStableNameExpr::Create(const ASTContext &Ctx,
+                                                   SourceLocation OpLoc,
+                                                   SourceLocation LParen,
+                                                   SourceLocation RParen,
+                                                   Expr *E) {
+  assert(E->isInstantiationDependent() &&
+         "Expr type only valid if the expr is dependent");
+  void *Mem = Ctx.Allocate(totalSizeToAlloc<Stmt *, TypeSourceInfo *>(1, 0),
+                           alignof(UniqueStableNameExpr));
+
+  QualType ResultTy = Ctx.getPointerType(Ctx.CharTy.withConst());
+  return new (Mem) UniqueStableNameExpr(OpLoc, LParen, RParen, ResultTy, E);
+}
+
+UniqueStableNameExpr *UniqueStableNameExpr::Create(const ASTContext &Ctx,
+                                                   SourceLocation OpLoc,
+                                                   SourceLocation LParen,
+                                                   SourceLocation RParen,
+                                                   TypeSourceInfo *TSI) {
+  void *Mem = Ctx.Allocate(totalSizeToAlloc<Stmt *, TypeSourceInfo *>(1, 0),
+                           alignof(UniqueStableNameExpr));
+  QualType ResultTy = Ctx.getPointerType(Ctx.CharTy.withConst());
+  return new (Mem) UniqueStableNameExpr(OpLoc, LParen, RParen, ResultTy, TSI);
+}
+
+UniqueStableNameExpr *UniqueStableNameExpr::CreateEmpty(const ASTContext &Ctx,
+                                                        bool IsExpr) {
+  void *Mem =
+      Ctx.Allocate(totalSizeToAlloc<Stmt *, TypeSourceInfo *>(IsExpr, !IsExpr),
+                   alignof(UniqueStableNameExpr));
+  QualType ResultTy = Ctx.getPointerType(Ctx.CharTy.withConst());
+  return new (Mem) UniqueStableNameExpr(EmptyShell(), ResultTy, IsExpr);
+}
+
+std::string UniqueStableNameExpr::ComputeName(ASTContext &Context) {
+  auto Callback = [](ASTContext &Ctx, const TagDecl *TD, raw_ostream &OS) {
+    if (!Ctx.IsSYCLKernelNamingDecl(TD))
+      return false;
+
+    // TODO: Can we put something here to make it clear this is a SYCL lambda?
```

Comment:
@schittir : This is what you said you would help with... I only did a 10 minute look into the mangling and found that this change worked for demangling, but please provide what you've discovered as the best way to insert something into the mangling such that it demangles.

Reply:
Hi Erich, 
If I understand this correctly, your current mangling approach is same as what I last mentioned i.e., `E5_` would demangle to `lambda5::`, where 5 is the stable lambda ordering number (which is WIP and you will substitute the real number later). However, in addition, you would like to produce a mangle pattern that has an explicit indicator that it is a sycl lambda, but this would be skipped or ignored by the demangler. Correct?

Reply:
Correct.  I was asking what you came up with during your analysis of the Itanium Mangling standard for this.  You'd mentioned a few months ago you had a way of mangling that would work and demangle, so I was asking you to share it.